src/stan_math_backend/Transform_Mir.ml

```diff
@@ -487,31 +466,70 @@ let trans_prog (p : Program.Typed.t) =
-        ; meta } in
-  let eval_udf_indexed_stmts (s : 'a Program.fun_def) =
-    if Set.mem possibly_recursive_fns (s.fdname, s.fdargs) then
-      {s with fdbody= Option.map ~f:map_stmt s.fdbody}
-    else s in
-  let p =
-    { p with
-      functions_block= List.map ~f:eval_udf_indexed_stmts p.functions_block }
-  in
+  let callgraph = String.Table.create () in
+  let eval_eigen_cycles fun_args calls (f : _ Program.fun_def) =
+    let open Expr.Fixed in
+    let rec is_potentially_recursive = function
+      | {pattern= Var name; _} -> Set.mem fun_args name
+      | {pattern= Indexed (e, _); _} -> is_potentially_recursive e
+      | {pattern= FunApp (StanLib (fname, _, _), e :: _); _} ->
+          Set.mem eigen_block_expr_fns fname && is_potentially_recursive e
+      | _ -> false in
+    let rec map_args name args =
+      let args = List.map ~f:rewrite_expr args in
+      let can_recurse, eval_args =
+        List.fold_map ~init:false
+          ~f:(fun is_rec e ->
+            if is_potentially_recursive e then
+              ( true
+              , {e with pattern= FunApp (StanLib ("eval", FnPlain, AoS), [e])}
+              )
+            else (is_rec, e) )
+          args in
+      if not can_recurse then args
+      else if name = f.fdname then eval_args
+      else
+        match Hashtbl.find callgraph name with
+        | Some nested when Hash_set.mem nested f.fdname -> eval_args
+        | Some nested ->
+            (* `calls` records all functions reachable from the current function *)
+            Hash_set.add calls name ;
+            Hash_set.iter nested ~f:(Hash_set.add calls) ;
+            args
+        | None -> Hash_set.add calls name ; args
+    and rewrite_expr = function
+      | {pattern= FunApp ((UserDefined (name, _) as kind), args); _} as e ->
+          {e with pattern= FunApp (kind, map_args name args)}
+      | e -> {e with pattern= Pattern.map rewrite_expr e.pattern} in
+    let rec rewrite_stmt s =
+      let open Stmt.Fixed in
+      match s with
+      | {pattern= Pattern.NRFunApp ((UserDefined (name, _) as kind), args); _}
+        as s ->
+          {s with pattern= NRFunApp (kind, map_args name args)}
+      | s -> {s with pattern= Pattern.map rewrite_expr rewrite_stmt s.pattern}
+    in
+    Program.map_fun_def rewrite_stmt f in
+  let break_cycles (Program.{fdname; fdargs; _} as fd) =
+    let fun_args =
+      List.filter_map fdargs ~f:(fun (_, n, t) ->
+          if UnsizedType.is_eigen_type t then Some n else None )
+      |> String.Set.of_list in
+    if Set.is_empty fun_args then fd
+    else
+      let calls =
+        Hashtbl.find_or_add callgraph fdname ~default:String.Hash_set.create
+      in
+      let fndef = eval_eigen_cycles fun_args calls fd in
+      (* update `callgraph` with the call paths going through the current function *)
+      Hashtbl.map_inplace callgraph ~f:(fun x ->
+          if Hash_set.mem x fdname then Hash_set.union calls x else x ) ;
+      fndef in
```

Comment:
Perhaps I'm missing something, but does this code still not have an order-dependence in the callgraph population? I don't see how it would have information "flow upward" if the final function in the functions block implicated prior ones, for example.

Also, not sure if this is relevant, but might as well ask: can you imagine a situation in which a function is recursive through the use of a higher-order function? Could that exhibit this poor eigen behavior?

Reply:
If this function could implicate a prior one then that prior function must already implicate this function. We eval()'ed all cycles that would go through this function so at the point where `callgraph` updates it's still cycle-free.

Good point about higher-order functions, this fails on master
```stan
functions {
  vector foo(real x, vector s, matrix y) {
    return ode_rk45(foo, [1]', 0.0, {1.0}, y[2:])[1];
  }
}
parameters {
  real x;
  vector[3] s;
  matrix[3,3] y;
}
transformed parameters {
   vector[3] z = foo(x,s,y);
}
```

Reply:
Assuming you can fix that in this framework, it would be good to add that as a test.

We should also probably add the small program you have in the original PR text as a regression test

Reply:
After staring at it a bit longer, I think I understand: we only need to insert an eval _somewhere_ in the chain to prevent the issue. So if we're looking at functions A, B, C, it doesn't matter if A doesn't know to eval calls to C, since when we get to C it will know it should.

This is actually even better than a solution which does what I thought it was, since it won't insert more than are strictly needed.